lib/advisor/core/questions/phrases_catalog.ex

```diff
@@ -1,52 +1,46 @@
 defmodule Advisor.Core.Questions.PhrasesCatalog do
-  alias Advisor.Core.Questions.PhraseKind
+  alias Advisor.Core.Questions.Phrase
 
   @path "lib/advisor/core/questions/questions.yml"
 
-  # TODO Maybe add a struct called `phrase`?
+  def find(ids) do
+    all()
+    |> flatten()
+    |> Enum.filter(fn(%Phrase{id: id}) -> id in ids end)
+  end
 
-  # TODO This feel brutally complicated... 
-  # ...and do I really need them all? Maybe I can push this into a spearet module too
+  # TODO Consider moving the entire loading, path joining shenaningsn out of here
   def all() do
     File.cwd!
     |> Path.join(@path)
-    |> all()
-  end
-  def all(path) do
-    path
     |> YamlElixir.read_from_file()
     |> extract()
   end
-  def from_data(yaml) do
-    yaml
-    |> YamlElixir.read_from_string()
-    |> extract()
-  end
 
-  # TODO: Gawd this is intricate and complicated...
   defp extract(yaml) do
     yaml
-    |> Enum.reduce(%{counter: 1, data: %{}}, fn({kind, phrases}, %{counter: counter, data: data}) ->
-      # TODO: Extract this function to its own module
-      kind = String.to_atom(kind)
-      questions = convert(phrases, kind, counter)
-      %{counter: counter + length(questions), data: Map.put(data, kind, questions)}
-    end)
-    |> Map.fetch!(:data) # Why do I just keep a few little chunks?!
+    |> zip()
+    |> Enum.with_index()
+    |> Enum.map(&to_phrase/1)
+    |> Enum.group_by(&kind/1, &(&1))
   end
 
-  def find(yaml, ids) do
-    yaml |> flatten() |> Enum.filter(fn(question) -> question.id in ids end)
-  end
-  def find(ids), do: all() |> find(ids)
+  defp kind(%Phrase{kind: kind}), do: kind
 
-  defp flatten(question_map), do: question_map |> Map.values |> List.flatten
+  defp zip(map) do
+    map
+    |> Enum.flat_map(fn({key, elements}) ->
+      Enum.map(elements, fn(v) -> {key, v} end)
+    end)
```

Comment:
What's `map` in this context? I also think it's interesting that we need to do a double map over the data structure to get what we want (`flat_map` -> `map`) 🤔

Reply:
I agree, a lot of maps going on here 🗺

Reply:
Maps over maps with maps!
the input has this shape from the YAML file:  `%{foo: ["phrase", "other"], bar: ["blue", "red"]}` and as it stands I need to create a `%Phrase{phrase: ..., kind: ..., id: ...}` struct. So the `flat_map` with `map` combination converts the structure into `[{"phrase", :foo}, {"other", :foo}, {"blue", :bar}, {"red", :bar}]` which is easier to process. Unless I make the change more invasive, I don't think I can remove that structure just yet.